frame/preimage/src/migration.rs

```diff
@@ -0,0 +1,212 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storage migrations for the preimage pallet.
+
+use super::*;
+#[cfg(feature = "try-runtime")]
+use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+use frame_support::{
+	storage_alias,
+	traits::{ConstU32, OnRuntimeUpgrade},
+};
+
+/// The log target.
+const TARGET: &'static str = "runtime::preimage::migration::v1";
+
+/// The original data layout of the preimage pallet without a specific version number.
+mod v0 {
+	use super::*;
+
+	#[derive(Clone, Eq, PartialEq, Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+	pub enum RequestStatus<AccountId, Balance> {
+		Unrequested(Option<(AccountId, Balance)>),
+		Requested(u32),
+	}
+
+	#[storage_alias]
+	pub type PreimageFor<T: Config> = StorageMap<
+		Pallet<T>,
+		Identity,
+		<T as frame_system::Config>::Hash,
+		BoundedVec<u8, ConstU32<MAX_SIZE>>,
+	>;
+
+	#[storage_alias]
+	pub type StatusFor<T: Config> = StorageMap<
+		Pallet<T>,
+		Identity,
+		<T as frame_system::Config>::Hash,
+		RequestStatus<<T as frame_system::Config>::AccountId, BalanceOf<T>>,
+	>;
+
+	#[cfg(feature = "try-runtime")]
+	pub fn image_count<T: Config>() -> Option<u32> {
+		// Use iter_values since otherwise it would indicate that a value cannot be decoded,
+		// possibly due to `MAX_SIZE` being too small.
+		let images = v0::PreimageFor::<T>::iter_values().count() as u32;
+		let status = v0::StatusFor::<T>::iter().count() as u32;
+
+		if images == status && images > 0 {
+			Some(images)
+		} else {
+			None
+		}
+	}
+}
+
+pub mod v1 {
+	use super::*;
+
+	/// Migration for moving preimages from a single map into into specific buckets.
+	///
+	/// Note: This needs to be run with the same hashing algorithm as before
+	/// since it is not re-hashing the preimages.
+	pub struct Migration<T>(sp_std::marker::PhantomData<T>);
```

Comment:
this migration wont be suitable for parachains that have reasonable use of the preimage pallet. probably a follow up we can delegate to the community is another migration which does this over multiple blocks.

Reply:
"Reasonable" use should be fine :)

If they're using it for megabytes of data then it's a little different, but are there any doing that?